### PR TITLE
Add `--module node18`

### LINKED ISSUE #60705: Add `--module node18`
Node v16 is EOL, and `--module node16` is an awkward target because it doesn't support import attributes, but requires import _assertions_ to import JSON. With #60534, `nodenext` will advance to support `require(esm)`, so we will want to have a module target that represents Node.js _beore_ `require(esm)` but _after_ import attributes.

`--module node18` will be identical to 5.7's `nodenext` (except perhaps for the implied `--target`), and `--module node16` will be deprecated in TypeScript 6.0.

### PR DESCRIPTION
**Pull request by andrewbranch**
Identical to `nodenext`, except implies `--target es2022` like `node16` does.

I validated the new baselines with

```
for f in tests/baselines/reference/*module=node18*; do
  [[ -e ${f/module=node18/module=node16} ]] && echo "$f" && diff "$f" "${f/module=node18/module=node16}";
done
```

and

```
for f in tests/baselines/reference/*module=node18*; do
  [[ -e ${f/module=node18/module=nodenext} ]] && echo "$f" && diff "$f" "${f/module=node18/module=nodenext}";
done
```

Closes #60705

### CODE REVIEW

File: tests/baselines/reference/nodeModulesJson(module=node16).errors.txt
Comment (andrewbranch) [inline tests/baselines/reference/nodeModulesJson(module=node16).errors.txt:1]: Compare this file...

File: tests/baselines/reference/nodeModulesJson(module=node18).errors.txt
Comment (andrewbranch) [inline tests/baselines/reference/nodeModulesJson(module=node18).errors.txt:1]: ...with this one to see the point of this PR

**Comment by jakebailey** [comment]
The example you gave has a diff like:

```diff
diff --git a/tests/baselines/reference/nodeModulesJson(module=node16).errors.txt b/tests/baselines/reference/nodeModulesJson(module=node18).errors.txt
index b3f36d73fa..033a1c9f7c 100644
--- a/tests/baselines/reference/nodeModulesJson(module=node16).errors.txt
+++ b/tests/baselines/reference/nodeModulesJson(module=node18).errors.txt
@@ -1,10 +1,9 @@
-/loosey.cts(1,36): error TS2823: Import attributes are only supported when the '--module' option is set to 'esnext', 'node18', 'nodenext', or 'preserve'.
+/loosey.cts(1,36): error TS2856: Import attributes are not allowed on statements that compile to CommonJS 'require' calls.
 /loosey.cts(6,9): error TS2339: Property 'default' does not exist on type '{ version: number; }'.
-/main.mts(5,36): error TS2823: Import attributes are only supported when the '--module' option is set to 'esnext', 'node18', 'nodenext', or 'preserve'.
-/main.mts(6,52): error TS2823: Import attributes are only supported when the '--module' option is set to 'esnext', 'node18', 'nodenext', or 'preserve'.
-/main.mts(8,10): error TS1544: Named imports from a JSON file into an ECMAScript module are not allowed when 'module' is set to 'Node16'.
-/main.mts(8,41): error TS2823: Import attributes are only supported when the '--module' option is set to 'esnext', 'node18', 'nodenext', or 'preserve'.
-/main.mts(9,42): error TS2823: Import attributes are only supported when the '--module' option is set to 'esnext', 'node18', 'nodenext', or 'preserve'.
+/main.mts(2,22): error TS1543: Importing a JSON file into an ECMAScript module requires a 'type: "json"' import attribute when 'module' is set to 'Node18'.
+/main.mts(3,19): error TS1543: Importing a JSON file into an ECMAScript module requires a 'type: "json"' import attribute when 'module' is set to 'Node18'.
+/main.mts(7,21): error TS1543: Importing a JSON file into an ECMAScript module requires a 'type: "json"' import attribute when 'module' is set to 'Node18'.
+/main.mts(8,10): error TS1544: Named imports from a JSON file into an ECMAScript module are not allowed when 'module' is set to 'Node18'.
 /main.mts(10,9): error TS2339: Property 'version' does not exist on type '{ default: { version: number; }; }'.
 
 
@@ -42,26 +41,24 @@
       "version": 1
     }
     
-==== /main.mts (6 errors) ====
+==== /main.mts (5 errors) ====
     import { oops } from "not.json"; // Ok
     import moreOops from "actually-json"; // Error in nodenext
+                         ~~~~~~~~~~~~~~~
+!!! error TS1543: Importing a JSON file into an ECMAScript module requires a 'type: "json"' import attribute when 'module' is set to 'Node18'.
     import typed from "actually-json/typed"; // Error in nodenext
+                      ~~~~~~~~~~~~~~~~~~~~~
+!!! error TS1543: Importing a JSON file into an ECMAScript module requires a 'type: "json"' import attribute when 'module' is set to 'Node18'.
     
     import config from "./config.json" with { type: "json" }; // Ok
-                                       ~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2823: Import attributes are only supported when the '--module' option is set to 'esnext', 'node18', 'nodenext', or 'preserve'.
     import { default as config1 } from "./config.json" with { type: "json" }; // Ok
-                                                       ~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2823: Import attributes are only supported when the '--module' option is set to 'esnext', 'node18', 'nodenext', or 'preserve'.
     import config2 from "./config.json"; // Error in nodenext, no attribute
+                        ~~~~~~~~~~~~~~~
+!!! error TS1543: Importing a JSON file into an ECMAScript module requires a 'type: "json"' import attribute when 'module' is set to 'Node18'.
     import { version } from "./config.json" with { type: "json" }; // Error, named import
              ~~~~~~~
-!!! error TS1544: Named imports from a JSON file into an ECMAScript module are not allowed when 'module' is set to 'Node16'.
-                                            ~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2823: Import attributes are only supported when the '--module' option is set to 'esnext', 'node18', 'nodenext', or 'preserve'.
+!!! error TS1544: Named imports from a JSON file into an ECMAScript module are not allowed when 'module' is set to 'Node18'.
     import * as config3 from "./config.json" with { type: "json" };
-                                             ~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2823: Import attributes are only supported when the '--module' option is set to 'esnext', 'node18', 'nodenext', or 'preserve'.
     config3.version; // Error
             ~~~~~~~
 !!! error TS2339: Property 'version' does not exist on type '{ default: { version: number; }; }'.
@@ -70,7 +67,7 @@
 ==== /loosey.cts (2 errors) ====
     import config from "./config.json" with { type: "json" }; // Error
                                        ~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2823: Import attributes are only supported when the '--module' option is set to 'esnext', 'node18', 'nodenext', or 'preserve'.
+!!! error TS2856: Import attributes are not allowed on statements that compile to CommonJS 'require' calls.
     import config2 from "./config.json"; // Ok
     import { version } from "./config.json"; // Ok
     import * as config3 from "./config.json";
```

Maybe this is preexisting, but with:

```diff
@@ -70,7 +67,7 @@
 ==== /loosey.cts (2 errors) ====
     import config from "./config.json" with { type: "json" }; // Error
                                        ~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2823: Import attributes are only supported when the '--module' option is set to 'esnext', 'node18', 'nodenext', or 'preserve'.
+!!! error TS2856: Import attributes are not allowed on statements that compile to CommonJS 'require' calls.
     import config2 from "./config.json"; // Ok
```

If someone wants to dual emit CJS/ESM but needs to write an attribute for JSON, are they out of luck? `require` allows JSON so I would have sort of expected us to say "okay" to this. But maybe it was already this restrictive?

**Comment by andrewbranch** [comment]
Yeah, that is preexisting, but a good point.

**Comment by jakebailey** [comment]
Well, you can't write `import =` if you want to be able to emit ESM, but you can of course always just read the file from the FS (ignoring potential emit path differences).

Just to be clear, is it preexisting as in "always", or preexisting in 5.7 specifically due to our added checks for attributes?

**Comment by andrewbranch** [comment]
That’s not true; `import =` compiles to `createRequire(import.meta.url)(...)`

This looks like a new problem _for us_ because of the new checks, but Node.js has enforced import assertions/attributes on JSON imports even while we have not, so we aren’t closing any doors with this. Anyone trying to do this previously has already had to work around it for runtime reasons, without help from TS.

**Comment by jakebailey** [comment]
Wow. I had no idea we did this:

```ts
import blah = require("react");

console.log(blah);
```

```js
import { createRequire as _createRequire } from "module";
const __require = _createRequire(import.meta.url);
const blah = __require("react");
console.log(blah);
```

File: tests/baselines/reference/transpile/Report an error when compiler-options module-kind is out-of-range.js [hunk 1]
```diff
@@ -1 +1,3 @@
+"use strict";
+Object.defineProperty(exports, "__esModule", { value: true });
```
Comment (jakebailey) [inline tests/baselines/reference/transpile/Report an error when compiler-options module-kind is out-of-range.js:2]: This seems like an interesting case; if the module kind is out of range, how are we ending up with CJS?

File: tests/cases/conformance/moduleResolution/packageJsonImportsExportsOptionCompat.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// @moduleResolution: classic, node, node16, nodenext, bundler
+// @moduleResolution: classic, node, node16,  nodenext, bundler
```
Comment (jakebailey) [inline tests/cases/conformance/moduleResolution/packageJsonImportsExportsOptionCompat.ts:1]: This seems like a typo (not that it matters)

**Review by jakebailey — APPROVED** [review]
Otherwise I think this is good.

**Pull request merged**